packages/entity-server/src/models/types.ts

```diff
@@ -6,12 +6,49 @@
 import { DatabaseModel, DatabaseType } from '@tupaia/database';
 import { ObjectLikeKeys, Flatten } from '../types';
 
-type PartialOrArray<T> = {
-  [field in keyof T]?: T[field] | T[field][];
+type FilterComparators = '!=' | 'ilike';
+
+type AdvancedFilterValue<T> = {
+  comparator: FilterComparators;
+  comparisonValue: T | T[];
+};
+
+/**
+ * Example:
+ * {
+ *    type: 'facility',
+ *    name: {
+ *        comparator: 'ilike',
+ *        comparisonValue: '%Pallet%'
+ *    }
+ * }
+ */
+type NormalCriteria<T> = {
+  [field in keyof T]?: T[field] | T[field][] | AdvancedFilterValue<T[field]>;
 };
 
-export type DbConditional<T> = PartialOrArray<Omit<T, ObjectLikeKeys<T>>> &
-  PartialOrArray<Flatten<Pick<T, ObjectLikeKeys<T>>, '->>'>>;
+export enum QueryConjunctions {
+  AND = '_and_',
+  OR = '_or_',
+  RAW = '_raw_',
+}
+
+/**
+ * Example:
+ * {
+ *    _and_: {
+ *        country_code: ['TO', 'KI']
+ *    }
+ * }
+ */
+type ConjunctionCriteria<T> = {
+  [field in QueryConjunctions]?: NormalCriteria<T>;
```

Comment:
```suggestion
  [field in QueryConjunctions]?: FilterCriteria<T>;
```
I _believe_ that we support recursion in this config, eg:
```
{
  name: 'TO'
  _or_: {
    type: 'facility',
    _or_: {
      parent_id: '1234',
      location_type: 'area'
    }  
  }
}
=>
WHERE code = 'TO' OR (type = 'facility' OR (parent_id = '1234' and location_type = 'area'))
```

Lucky for us, type also supports all kind of recursive insanity 🎉